### PR TITLE
[Feature 3] Upgrade System

### PR DESCRIPTION
**Pull request by RitatataVan**
1. Updated upgrade functionality and link to MatcherUI 
2. Added java docs for all files under package upgrade

### CODE REVIEW

File: src/main/java/upgrade/UpgradePresenter.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package upgrade;
+
+public class UpgradePresenter implements UpgradeOutputBoundary {
+
+    private String currentEmail;
+    private boolean finishSubscribed;
+
+    public UpgradePresenter(String currentEmail) {
+        this.currentEmail = currentEmail;
+        this.finishSubscribed = false;
```
Comment (ShuHaoLin2002) [inline src/main/java/upgrade/UpgradePresenter.java:10]: may be redundant as the default value of Boolean is false.

**Review by ShuHaoLin2002 — COMMENTED** [review]
1. The possibility to include request and response models to bundle data in controller and presenter to enhance coding structure. 
2. setUserRepoManager and setOutputBoundary in UpgradeManager and UpgradeInteractor may be replaced using their Constructor. 
3. May be able to combine LoginInteractor and LoginManager as a single class.

File: src/main/java/upgrade/UpgradePresenter.java (hunk 1 above)
Comment (RitatataVan) [reply src/main/java/upgrade/UpgradePresenter.java:10]: I think this should be included in the constructor, though the default value is false

**Comment by RitatataVan** [comment]
1 and 2 make sense to me. I'll refactor them later. btw I think I include UpgradePresenterViewModel as the response model.
For 3, do you mean UpgradeInteractor and UpgradeManager? I don't have LoginInteractor and LoginManager in my branch.

File: src/main/java/upgrade/UpgradePresenter.java (hunk 1 above)
Comment (cyltaracai) [reply src/main/java/upgrade/UpgradePresenter.java:10]: Shuhao is right about that finishSubscribed is not need in the constructor to initialize it. But for readability, it's good to include finishSubscribed in the constructor. Leaving it like this is fine.

**Review by cyltaracai — APPROVED** [review]
I reviewed how the clean architecture is implemented in this feature. No clean architecture violation is identified. Great Work.
There is a bug in implementing Singleton pattern for UpgradeManager.  To complete the pattern, need to make its constructor private. See the detailed comment for more discussions about Singleton pattern.

File: src/main/java/upgrade/UpgradePresenter.java (hunk 1 above)
Comment (MacroSony) [reply src/main/java/upgrade/UpgradePresenter.java:10]: Agree  with @ShuHaoLin2002  and @cyltaracai

**Review by MacroSony — COMMENTED** [review]
Not much to comment on, great work

**Review by cyltaracai — APPROVED** [review]
All classes still need java docs and comments.  But I'll approve this pull request so that we can start working on merging features. Please add java docs and comments later.  Thank you

**Review by cyltaracai — APPROVED** [review]
Looks good, except the autograding.

**Review by cyltaracai — DISMISSED** [review]

**Review by MacroSony — DISMISSED** [review]
Great Work

**Review by ShuHaoLin2002 — APPROVED** [review]

File: src/main/java/upgrade/UpgradeManager.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    private static UpgradeManager instance;
+
+    public static UpgradeManager instance() {
+        synchronized (UpgradeManager.class) {
+            if (instance == null) {
+                instance = new UpgradeManager();
+            }
+        }
+        return instance;
+    }
```
[the remark this answers is not in the document]
Comment (RitatataVan) [reply src/main/java/upgrade/UpgradeManager.java]: Has revised to normal constructor

**Comment by RitatataVan** [comment]
The failure in autograding has been resolved in the branch "Connecting features 6, 7, and 3 to work simultaneously".

**Pull request merged**